File: Day7.py
import itertools


def get_file_data(file_name):
    f = open(file_name)
    data = []
    for line in f:
        data.append(line.rstrip())
    f.close()
    return data


def evaluate_operation(operation):
    index = 1
    current_result = 0
    while index < len(operation):
        if index == 1:
            if operation[index] == "+":
                current_result += (operation[index-1] + operation[index+1])
            if operation[index] == "*":
                current_result += (operation[index-1] * operation[index+1])
            if operation[index] == "||":
                current_result += int((str(operation[index-1]) + str(operation[index+1])))
        else:
            if operation[index] == "+":
                current_result += operation[index+1]
            if operation[index] == "*":
                current_result *= operation[index+1]
            if operation[index] == "||":
                current_result = int(str(current_result) + str(operation[index+1]))
        index += 2
    return current_result


file_data = get_file_data("input_file")
results = []
operands = []
operator_list_part_one = []
operator_list_part_two = []

for line in file_data:
    number = line.split(":")[0]
    results.append(int(number))
    one_set_of_operands = line.split(":")[1].split(" ")
    one_set_of_operands.pop(0)
    for i in range(len(one_set_of_operands)):
        one_set_of_operands[i] = int(one_set_of_operands[i])
    operands.append(one_set_of_operands)

for operand in operands:
    combos = []
    for combo in itertools.product(("+", "*"), repeat=len(operand)-1):
        combos.append(combo)
    operator_list_part_one.append(combos)

    combos = []
    for combo in itertools.product(("+", "*", "||"), repeat=len(operand)-1):
        combos.append(combo)
    operator_list_part_two.append(combos)

all_combinations_part_one = []
all_combinations_part_two = []
for i in range(len(operands)):
    result = results[i]
    for operators in operator_list_part_one[i]:
        unique_operation = operands[i].copy()
        index = 1
        for o in operators:
            unique_operation.insert(index, o)
            index += 2
        all_combinations_part_one.append((unique_operation, result))
    for operators in operator_list_part_two[i]:
        unique_operation = operands[i].copy()
        index = 1
        for o in operators:
            unique_operation.insert(index, o)
            index += 2
        all_combinations_part_two.append((unique_operation, result))

good_results = set()

for item in all_combinations_part_one:
    result = evaluate_operation(item[0])
    if result == item[1]:
        good_results.add(result)

print("Part one answer:", sum(good_results))
good_results = set()

for item in all_combinations_part_two:
    result = evaluate_operation(item[0])
    if result == item[1]:
        good_results.add(result)

print("Part two answer:", sum(good_results))

File: Day7.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

public class Day7 {
    public static void main(String[] args) {
        long partOneAnswer = 0;
        long partTwoAnswer = 0;
        ArrayList<String> fileData = getFileData("src/InputFile");
        ArrayList<EquationChecker> all_equations = new ArrayList<EquationChecker>();
        for (String line : fileData) {
            long r = Long.parseLong(line.split(":")[0]);
            ArrayList<Integer> convertedOperands = new ArrayList<Integer>();
            String[] operands = line.split(": ")[1].split(" ");
            for (String o : operands) {
                convertedOperands.add(Integer.parseInt(o));
            }
            EquationChecker e = new EquationChecker(r, convertedOperands, false);
            if (e.isValidEquation()) partOneAnswer += e.getResult();
            EquationChecker e2 = new EquationChecker(r, convertedOperands, true);
            if (e2.isValidEquation()) partTwoAnswer += e2.getResult();
            all_equations.add(e);
        }

        System.out.println("Part one answer: " + partOneAnswer);
        System.out.println("Part two answer: " + partTwoAnswer);
    }

    public static ArrayList<String> getFileData(String fileName) {
        ArrayList<String> fileData = new ArrayList<String>();
        try {
            File f = new File(fileName);
            Scanner s = new Scanner(f);
            while (s.hasNextLine()) {
                String line = s.nextLine();
                if (!line.equals(""))
                    fileData.add(line);
            }
            return fileData;
        }
        catch (FileNotFoundException e) {
            return fileData;
        }
    }
}

class EquationChecker {
    private long result;
    private ArrayList<Integer> operands;
    private ArrayList<String> operatorCombinations;
    private ArrayList<ArrayList<Integer>> allEquations;
    private boolean validEquation;

    public EquationChecker(long result, ArrayList<Integer> operands, boolean partTwo) {
        this.result = result;
        this.operands = operands;
        String operatorsString = "01";
        if (partTwo) operatorsString = "012";
        operatorCombinations = generateCombinations(operatorsString, operands.size()-1);
        allEquations = new ArrayList<ArrayList<Integer>>();
        for (String operators : operatorCombinations) {
            int index = 1;
            ArrayList<Integer> equation = new ArrayList<Integer>();
            for (int n : operands) {
                equation.add(n);
            }
            for (int i = 0; i < operators.length(); i++) {
                int operator = Integer.parseInt(operators.substring(i, i+1));
                equation.add(index, operator);
                index += 2;
            }
            allEquations.add(equation);
        }

        for (ArrayList<Integer> equation : allEquations) {
            evaluateEquation(equation);
        }
    }

    private long evaluateEquation(ArrayList<Integer> equation) {
        long result = 0;
        for (int i = 1; i < equation.size(); i += 2) {
            int operator = equation.get(i);
            if (i == 1) {
                if (operator == 0) result += equation.get(0) + equation.get(i+1);
                if (operator == 1) result += (long) equation.get(0) * equation.get(i+1);
                if (operator == 2) result += Long.parseLong(("" + equation.get(0) + equation.get(i+1)));
            }
            else {
                if (operator == 0) result += equation.get(i+1);
                if (operator == 1) result *= equation.get(i+1);
                if (operator == 2) result = Long.parseLong("" + result + equation.get(i+1));
            }
        }
        if (result == this.result) validEquation = true;
        return result;
    }

    public String toString() {
        return "\nDesired result: " + result + "\nOperands: " + operands + "\n" + "Valid: " + validEquation;
    }

    private ArrayList<String> generateCombinations(String input, int length) {
        ArrayList<String> result = new ArrayList<>();
        generateCombinationsHelper(input, "", length, result);
        return result;
    }

    private void generateCombinationsHelper(String input, String current, int length, ArrayList<String> result) {
        if (current.length() == length) {
            result.add(current);
            return;
        }
        for (int i = 0; i < input.length(); i++) {
            generateCombinationsHelper(input, current + input.charAt(i), length, result);
        }
    }

    public boolean isValidEquation() {
        return validEquation;
    }

    public long getResult() {
        return result;
    }
}
